Add `RackAwareRoundRobinPolicy` for host selection

sylwiaszunejko opened the pull request:
Added `RackAwareRoundRobinPolicy` to prefer replicas in local rack while picking a coordinator. Similar to `DCAwareRoundRobinPolicy` but with extra option.

Tests are added.

Fixes: #325

sylwiaszunejko commented:
@Lorak-mmk @fruch ping

fruch commented on cassandra/policies.py:
```diff
@@ -316,6 +327,138 @@ def on_add(self, host):
+                yield host
+
+    def on_up(self, host):
+        # not worrying about threads because this will happen during
```
this kind of comment does make me worry :)

in general sense I think those kind of statements are better to backed up with specific tests (unit, integration) to "prove" them.

fruch commented on cassandra/policies.py:
```diff
@@ -316,6 +327,138 @@ def on_add(self, host):
+
+        local_live = tuple(host for (datacenter, rack), hosts in self._live_hosts.items() if datacenter == self.local_dc and rack != self.local_rack for host in hosts)
+        pos = (pos % len(local_live)) if local_live else 0
+        for host in islice(cycle(local_live), pos, pos + len(local_live)):
```
this combo of islice + cycle, might need a bit of description with a comment.

it's not clear to me why this positional work is needed, 

might be a clear why to write it down...

fruch commented on cassandra/policies.py:
```diff
@@ -316,6 +327,138 @@ def on_add(self, host):
+        return host.datacenter or self.local_dc
+
+    def populate(self, cluster, hosts):
+        for (dc, rack), dc_hosts in groupby(hosts, lambda h: (self._dc(h), self._rack(h))):
```
nitpick: `s/h/host/` a bit more readable

fruch commented on cassandra/policies.py:
```diff
@@ -316,6 +327,138 @@ def on_add(self, host):
+
+    def populate(self, cluster, hosts):
+        for (dc, rack), dc_hosts in groupby(hosts, lambda h: (self._dc(h), self._rack(h))):
+            self._live_hosts[(dc, rack)] = tuple(set(dc_hosts))
```
can we get to situation we have duplicates hosts, passed to this function ? that a `set` is needed ?

sylwiaszunejko replied:
I guess so, because every other policy uses `set`: 
https://github.com/scylladb/python-driver/blob/master/cassandra/policies.py#L164
https://github.com/scylladb/python-driver/blob/master/cassandra/policies.py#L240
https://github.com/scylladb/python-driver/blob/master/cassandra/policies.py#L451

fruch replied:
o.k., I was just wondering how we get to that situation.

fruch commented on cassandra/policies.py:
```diff
@@ -316,6 +327,138 @@ def on_add(self, host):
+                endpoint
+                for endpoint in cluster.endpoints_resolved]
+
+        self._position = randint(0, len(hosts) - 1) if hosts else 0
```
I think a few word regarding this internal implementation, and what this randomness is for, could be helpful.

I couldn't figure it from the code

fruch replied:
o.k. would still recommend to document it, even if upstream didn't

fruch commented on tests/unit/test_policies.py:
```diff
@@ -520,6 +520,382 @@ def test_default_dc(self):
+        self.assertEqual(qplan, [])
+
+    def test_no_nodes(self):
+        """
```
I think all of the tests, should have a __docstring__ stating the goals.

putting the goal inside the names of the test might be tricky a bit.

fruch replied:
now I've noticed (by comment mentioning datastax issue), that it's mostly copied test code.

if it's not massively different, maybe subclassing the test class would work a bit better ?

fruch reviewed (commented):
- few clarifications, and a bit more documenting the internal implementation
- I think adding integration test for this, should be very helpful, also as documentation on how to use it
- and documentation, the docstring of the class isn't enough, we should have an example how a user should be expected to set it up.  (in the java implementation was all bunch of hidden assumptions, and it took us quite time to figure why c-s wasn't working as expected), in this case we don't have a stress tool readily available to test the same in SCT, so would recommend investing documentation a bit more.

fruch commented:
as for the integration test with multi dc and multi-racks, I would recommend looking into similar test in dtest, and clone the needed ideas/helper functions from there.

sylwiaszunejko commented on cassandra/metadata.py:
```diff
@@ -3440,7 +3440,7 @@ def group_keys_by_replica(session, keyspace, table, keys):
         # First check if there are local replicas
         valid_replicas = [host for host in all_replicas if
-                          host.is_up and distance(host) == HostDistance.LOCAL]
+                          host.is_up and (distance(host) == HostDistance.LOCAL or distance(host) == HostDistance.LOCAL_RACK)]
```
This function consider replicas valid if they are in the same datacenter, if using `RackAwareRoundRobinPolicy` both hosts with the distance LOCAL and LOCAL_RACK are in the same datacenter.
To be honest I do not really get the purpose of `group_keys_by_replica` function, but I just didn't want to mess with its behavior

fruch commented:
> Maybe this new policy should be the default? I think you need to modify `default_lbp_factory` function in cluster.py to do this (+ comment in `class ExecutionProfile`'s `load_balancing_policy` attribute and perhaps some other comments). @mykaul wdyt?

what's the reason ? i.e. it would be doing the exact same thing, since no parameters are passed to it.


> In general please go over all occurences of `DCAwareRoundRobinPolicy` in the driver code (and docs etc) and see if they need to be replaced or complemented by `RackAwareRoundRobinPolicy`

again, don't know about replacing, but each occurence need to be understood, and figure if there's something todo with it.

fruch commented on cassandra/policies.py:
```diff
@@ -316,6 +327,138 @@ def on_add(self, host):
+    def on_up(self, host):
+        # not worrying about threads because this will happen during
+        # control connection startup/refresh
+        if not self.local_rack and host.rack:
+            if host.endpoint in self._endpoints:
+                self.local_rack = host.rack
+                self.local_dc = host.datacenter
+                log.info("Using rack '%s' and datacenter '%s' for RackAwareRoundRobinPolicy (via host '%s'); "
+                         "if incorrect, please specify a local_dc to the constructor, "
+                         "or limit contact points to local cluster nodes" %
+                         (self.local_rack, self.local_dc, host.endpoint))
+                del self._endpoints
```
I would argue, that this type of validation should happen on __init__

the case we might need to handle is hosts which don't match, cause that should be o.k. to pass all of the nodes as contact points

Lorak-mmk replied:
By **init** you mean `__init__` method of `RackAwareLoadBalancingPolicy`? If so, we are in agreement because that is what I meant by constructor.

fruch replied:
yes, look like I didn't read to the very end... my bad.

fruch commented on cassandra/policies.py:
```diff
@@ -316,6 +327,138 @@ def on_add(self, host):
+            if host.endpoint in self._endpoints:
+                self.local_rack = host.rack
+                self.local_dc = host.datacenter
+                log.info("Using rack '%s' and datacenter '%s' for RackAwareRoundRobinPolicy (via host '%s'); "
```
also, the text sounds like a warning, but the level is info.

I would rather we don't have warning like that, but errors when misconfigured

Lorak-mmk replied:
I don't think it's a warning. From what I understand there are 2 valid ways to use this policy: you can explicitly pass DC and Rack to policy, or you can skip it in which case it will be taken from one of the contact points.
So this is info log about which route was taken.

sylwiaszunejko replied:
Exactly

fruch replied:
What is the point of using this class if not passing in a rack ? (It's the same with the DC one,  but I don't think we should change that)

Letting the driver randomly set rack or data center, would be confusing to users.

every time they connect they might get different results, i.e. latency and such, cause different DC or different RACK was selected.

or I don't fully understand what those fallbacks are for...

sylwiaszunejko replied:
I made it that way due to how it looks like in DC aware policy, but we can decide to take another road.

fruch replied:
I'm just trying to assume to role of a user here, and suggest we put guardrails to begin with (no need to wait for people to get bitten to put them in :))

but anyhow it's a bit of a guess work, until we'll have contact with some who actually need it

Lorak-mmk replied:
To add some data points:
Rust Driver requires user to set dc and/or rack explicitly if they are to be preferred, there is no mechanism  to fetch it from contact points.
Java Driver 4.x seems to also require user to explicitly pass DC / Rack to enable DC / Rack awareness
Java Driver 3.x has similar mechanism to this PR

4.x and Rust don't have such mechanisms, so I think we should follow their lead

Lorak-mmk commented on cassandra/cluster.py:
```diff
@@ -622,6 +623,11 @@ class Cluster(object):
     surrounding unintuitive default load-balancing policy behavior.
+
+    Note: When using :class:`.RackAwareLoadBalancingPolicy` with no explicit
+    local_dc and local_rack set (as is the default), the DC is chosen from an arbitrary
```
Why did this note disappear in new version?

sylwiaszunejko replied:
Now local_dc and local_rack have to be explicit, so there is no choosing dc or rack from arbitrary host

Lorak-mmk commented on cassandra/policies.py:
```diff
@@ -316,6 +327,134 @@ def on_add(self, host):
+        with self._hosts_lock:
+            current_hosts = self._live_hosts.get((dc, rack), ())
+            if host in current_hosts:
+                hosts = tuple(h for h in current_hosts if h != host)
+                if hosts:
+                    self._live_hosts[(dc, rack)] = hosts
```
If dict stores lists then `list.remove` could be used here.

Lorak-mmk commented on cassandra/policies.py:
```diff
@@ -316,6 +327,134 @@ def on_add(self, host):
+                else:
+                    del self._live_hosts[(dc, rack)]
+            current_hosts_dc = self._dc_live_hosts.get(dc, ())
+            if host in current_hosts_dc:
+                hosts = tuple(h for h in current_hosts_dc if h != host)
+                if hosts:
+                    self._dc_live_hosts[dc] = hosts
+                else:
+                    del self._dc_live_hosts[dc]
```
Why is there no `del` branch in the previous fragment (handling `self._live_hosts`)?

sylwiaszunejko replied:
There is one in line 444

Lorak-mmk commented on tests/unit/test_policies.py:
```diff
@@ -520,6 +578,31 @@ def test_default_dc(self):
         policy.on_add(host_remote)
         self.assertFalse(policy.local_dc)
 
+class RackAwareRoundRobinPolicyTest(LocalAwareRoundRobinPolicyBase):
+
+    def test_no_remote(self):
+        return super().test_no_remote(RackAwareRoundRobinPolicy, ("dc1", "rack1"))
+
+    def test_with_remotes(self):
+        return super().test_with_remotes(RackAwareRoundRobinPolicy, ("dc1", "rack1"))
+
+    def test_get_distance(self):
+        return super().test_get_distance(RackAwareRoundRobinPolicy, ("dc1", "rack1"), HostDistance.LOCAL_RACK)
+
+    def test_status_updates(self):
+        return super().test_status_updates(RackAwareRoundRobinPolicy, ("dc1", "rack1"))
+
+    def test_modification_during_generation(self):
+        return super().test_modification_during_generation(RackAwareRoundRobinPolicy, ("dc1", "rack1"))
+
+    def test_no_live_nodes(self):
+        return super().test_no_live_nodes(RackAwareRoundRobinPolicy, ("dc1", "rack1"))
+
+    def test_no_nodes(self):
+        return super().test_no_live_nodes(RackAwareRoundRobinPolicy, ("dc1", "rack1"))
+
+
+
```
Yes, I think it's not needed with pytest, is it correct @fruch ?

fruch replied:
I think you can use it in tandem, but not sure
And if not, I think it's o.k. not to use the builtin unittest

We are running tests only with pytest, all around. (And if someone doesn't, it's about time he would)

sylwiaszunejko commented on cassandra/policies.py:
```diff
@@ -316,6 +327,134 @@ def on_add(self, host):
+            if not dc_hosts:
+                return HostDistance.IGNORED
+
+            if host in list(dc_hosts)[:self.used_hosts_per_remote_dc]:
```
Because every other policy uses set:
https://github.com/scylladb/python-driver/blob/master/cassandra/policies.py#L164
https://github.com/scylladb/python-driver/blob/master/cassandra/policies.py#L240
https://github.com/scylladb/python-driver/blob/master/cassandra/policies.py#L451
I just wanted to be in line with other policies

Lorak-mmk replied:
Imo we should understand why they are using set and not just following it.
Either there is a good reason, which should be documented, or there is not - then we could change it here and maybe in other policies.

sylwiaszunejko replied:
I believe there is no good reason

sylwiaszunejko commented:
v2:
* `@pytest.mark.parametrize` instead of subclasses in tests
* use `list` instead of `tuple(set())` in `populate`

sylwiaszunejko commented:
> > Maybe this new policy should be the default? I think you need to modify `default_lbp_factory` function in cluster.py to do this (+ comment in `class ExecutionProfile`'s `load_balancing_policy` attribute and perhaps some other comments). @mykaul wdyt?
> > In general please go over all occurences of `DCAwareRoundRobinPolicy` in the driver code (and docs etc) and see if they need to be replaced or complemented by `RackAwareRoundRobinPolicy`
> 
> Yes, I believe it should be the default. What is the default in other drivers?

In Rust driver rack awareness is a part of default load balancing policy, there is a field for preferred rack, if it is set rack awareness is used if not it is not used https://github.com/scylladb/scylla-rust-driver/blob/main/docs/source/load-balancing/default-policy.md#preferences
In java-driver 4.x the rack-awareness feature is optional and to enable it the local rack should
be supplied through the configuration. The feature is disabled by default and unlike the local datacenter it will not be implicitly fetched from the provided contact points. https://github.com/scylladb/java-driver/tree/scylla-4.x/manual/core/load_balancing#rack-awareness

mykaul commented:
> In Rust driver rack awareness is a part of default load balancing policy, there is a field for preferred rack, if it is set rack awareness is used if not it is not used https://github.com/scylladb/scylla-rust-driver/blob/main/docs/source/load-balancing/default-policy.md#preferences

This sounds like what we should support as well.

sylwiaszunejko commented:
> > In Rust driver rack awareness is a part of default load balancing policy, there is a field for preferred rack, if it is set rack awareness is used if not it is not used https://github.com/scylladb/scylla-rust-driver/blob/main/docs/source/load-balancing/default-policy.md#preferences
> 
> This sounds like what we should support as well.

We can pass `RackAwareRoundRobinPolicy` as a child policy in `TokenAwarePolicy`, isn't that an equivalent of specifying field? In Rust driver DC awareness is handled by a field and in python-driver by a child policy. 

If we would want to replace all occurrences of `DCAware` with `RackAware` it would not be obvious, in this function for example: https://github.com/scylladb/python-driver/blob/master/cassandra/cluster.py#L240-#L243 it will hard to use Rack aware instead of DC aware policy, because **to be consistent with rust driver and java driver 4.x** we decided that local_dc and local_rack have to be specified.

Lorak-mmk commented:
Now that we decided to not do implicit dc / rack it makes no sense imo to make this policy default. Because without specifying dc / rack it won't do anything, and as @sylwiaszunejko mentioned specifying it is equivalent to using it as child policy to TokenAware.

sylwiaszunejko commented:
I added some kind of pattern for integration tests, but I am not really sure how the tests should look like. I could use some help with that @Lorak-mmk @fruch

fruch commented on tests/integration/standard/test_rack_aware_policy.py:
```diff
@@ -0,0 +1,48 @@
+            response = self.session.execute('SELECT * from system.local')
+            queried_hosts.update(response.response_future.attempted_hosts)
+        queried_hosts = set(host.address for host in queried_hosts)
+        # Performe some checks
```
I think we can use a similar approach as in shard aware tests, and use tracing to validate requests arrived at the correct nodes

fruch commented on tests/integration/standard/test_rack_aware_policy.py:
```diff
@@ -0,0 +1,48 @@
+        self.cluster = Cluster(contact_points=["127.0.0.1", "127.0.0.2", "127.0.0.3", "127.0.0.4", "127.0.0.5", "127.0.0.6"], protocol_version=PROTOCOL_VERSION,
+                              load_balancing_policy=RackAwareRoundRobinPolicy("WRONG_DC1", "RAC1", used_hosts_per_remote_dc=0),
+                              reconnection_policy=ConstantReconnectionPolicy(1))
+        self.assertRaises(NoHostAvailable, self.cluster.connect)
```
I think this can be covered in unittest, no point of spinning up a cluster for such check

sylwiaszunejko replied:
moved it to unittests

fruch commented on tests/integration/standard/test_rack_aware_policy.py:
```diff
@@ -0,0 +1,48 @@
+class RackAwareRoundRobinPolicyTests(unittest.TestCase):
+    @classmethod
+    def setup_class(cls):
+        cls.cluster = Cluster(contact_points=["127.0.0.1", "127.0.0.2", "127.0.0.3", "127.0.0.4", "127.0.0.5", "127.0.0.6"], protocol_version=PROTOCOL_VERSION,
```
Don't use hardcoded addresses 

Ccm is used here we can extract the addresses from ccm cluster, as we do in dtest (and probably in driver tests as well)


It's something like that (writing it from memory)
```python
[node.address() for node in ccm_cluster.nodelist()]
```

sylwiaszunejko commented:
I am not sure what is going on with the CI. I cannot download full logs it keeps failing every time. Locally all tests pass

sylwiaszunejko commented:
@Lorak-mmk @fruch ping

Lorak-mmk commented on cassandra/policies.py:
```diff
@@ -316,6 +327,118 @@ def on_add(self, host):
+    def on_down(self, host):
+        dc = self._dc(host)
+        rack = self._rack(host)
+        with self._hosts_lock:
+            self._live_hosts[(dc, rack)].remove(host)
+            self._dc_live_hosts[dc].remove(host)
```
This version will result in empty lists (when all hosts in given dc / rack go down), previous version removed the key from dict in that case.
Is that going to cause problems? Is there some code that assumes values to be non-empty?

Lorak-mmk replied:
I wanted to say that it has the same problem as `on_up` but there should never be `on_down` for a host that the policy doesn't know about, right?

sylwiaszunejko replied:
I added back removing key form the dict

> I wanted to say that it has the same problem as on_up but there should never be on_down for a host that the policy doesn't know about, right?

yes

Lorak-mmk commented on cassandra/policies.py:
```diff
@@ -316,6 +327,118 @@ def on_add(self, host):
+        dc = self._dc(host)
+        rack = self._rack(host)
+        with self._hosts_lock:
+            self._live_hosts[(dc, rack)].append(host)
+            self._dc_live_hosts[dc].append(host)
```
This will crash if there is no such dc / rack in the dict.
It's sad that our tests don't detect this. Can you write a unit test that does?

To fix this, you can used `https://docs.python.org/3/library/stdtypes.html#dict.setdefault`:
`self._live_hosts.setdefault((dc, rack), []).append(host)` and similarly for `_dc_live_hosts`

sylwiaszunejko replied:
Tests didn't detect this because after changing unit test to parametrized one and removing `unittest.TestCase` the test was not executed. I've realized this now when trying to run only one test utilizing `on_add` to debug why it doesn't detect this bug.

sylwiaszunejko replied:
I had to rename class to start with "Test*" to be detected without `unittest.TestCase`

Lorak-mmk commented on cassandra/policies.py:
```diff
@@ -316,6 +327,120 @@ def on_add(self, host):
+
+        # the dict can change, so get candidate DCs iterating over keys of a copy
+        other_dcs = [dc for dc in self._dc_live_hosts.copy().keys() if dc != self.local_dc]
+        for dc in other_dcs:
```
I think you could write `for (dc, remote_live) in self._dc_live_hosts.copy() if dc != self.local_dc` (or similar, I didn't test it). That way previous line becomes unnecessary (which saves an allocation), and you don't nee to use `get` with default in next line.

Lorak-mmk commented on cassandra/policies.py:
```diff
@@ -316,6 +327,120 @@ def on_add(self, host):
+        with self._hosts_lock:
+            self._live_hosts[(dc, rack)].remove(host)
+            self._dc_live_hosts[dc].remove(host)
+            if not self._dc_live_hosts[dc]:
+                del self._dc_live_hosts[dc]
```
Don't you need to remove from `self._live_hosts` too?

Lorak-mmk commented on tests/integration/standard/test_rack_aware_policy.py:
```diff
@@ -0,0 +1,86 @@
+            bound = prepared.bind((i, i%5, i%2))
+            session.execute(bound)
+
+    def test_rack_aware(self):
+        prepared = self.session.prepare(
+            """
+            SELECT pk, ck, v FROM test1.table1 WHERE pk = ?
+            """)
+
+        bound = prepared.bind([(1)])
+        results = self.session.execute(bound, trace=True)
+        self.assertEqual(results, [(1, 1, 1)])
+        coordinator = str(results.response_future.coordinator_host.endpoint)
+        self.assertTrue(coordinator in set(["127.0.0.1:9042", "127.0.0.2:9042"]))
+
+        self.node2.stop(wait_other_notice=True, gently=True)
+
+        bound = prepared.bind([(2)])
+        results = self.session.execute(bound, trace=True)
+        self.assertEqual(results, [(2, 2, 0)])
+        coordinator =str(results.response_future.coordinator_host.endpoint)
+        self.assertEqual(coordinator, "127.0.0.1:9042")
+
+        self.node1.stop(wait_other_notice=True, gently=True)
+
+        bound = prepared.bind([(3)])
+        results = self.session.execute(bound, trace=True)
+        self.assertEqual(results, [(3, 3, 1)])
+        coordinator = str(results.response_future.coordinator_host.endpoint)
+        self.assertTrue(coordinator in set(["127.0.0.3:9042", "127.0.0.4:9042"]))
```
The test should probably execute each part a few times in a loop.

sylwiaszunejko replied:
10 times will be enough?

Lorak-mmk replied:
Yes

Lorak-mmk commented on tests/unit/test_policies.py:
```diff
@@ -180,130 +181,167 @@ def test_no_live_nodes(self):
         qplan = list(policy.make_query_plan())
-        self.assertEqual(set(qplan[:2]), set([hosts[1], new_local_host]))
-        self.assertEqual(set(qplan[2:]), set([hosts[3], new_remote_host]))
+        print("QPLAN: ", qplan)
```
Probably better to use logging instead of printing (or just remove this).

Lorak-mmk commented on tests/unit/test_policies.py:
```diff
@@ -460,27 +498,37 @@ def test_no_live_nodes(self):
+        qplan = list(policy.make_query_plan())
+        assert len(qplan) == 0
+
+    def test_default_dc(self, policy_specialization, constructor_args):
         host_local = Host(DefaultEndPoint(1), SimpleConvictionPolicy, 'local')
         host_remote = Host(DefaultEndPoint(2), SimpleConvictionPolicy, 'remote')
         host_none = Host(DefaultEndPoint(1), SimpleConvictionPolicy)
```
This test only uses `DCAwareRoundRobinPolicy`, ignores specialization, so it should probably be outside the class.

Lorak-mmk reviewed (changes requested):
I left few more comments, the main will be imo fine after you fix them.
Regarding unit tests: I can't say I love that many tests need conditions to have different branches for both policies, but I don't see how to write them better in this regard. Also it's probably still better than copy pasting the whole test class (because with any changes there is a high risk of 2 classes getting out of sync).

@fruch wdyt?

sylwiaszunejko commented:
I am still fighting with the CI, this part of the code in `policies.py` seems to be problematic:

```python
for host in islice(cycle(local_rack_live), pos, pos + len(local_rack_live)):
    yield host
```

Even though `local_rack_live = [<Host: 127.0.0.2:9042 DC1>, <Host: 127.0.0.1:9042 DC1>]` first host is yield two times, and because neither of these cases results in a successful read (node2 is stopped), driver uses host with not local rack. 
I am not sure why that happens.

sylwiaszunejko commented:
> I am still fighting with the CI, this part of the code in `policies.py` seems to be problematic:
> 
> ```python
> for host in islice(cycle(local_rack_live), pos, pos + len(local_rack_live)):
>     yield host
> ```
> 
> Even though `local_rack_live = [<Host: 127.0.0.2:9042 DC1>, <Host: 127.0.0.1:9042 DC1>]` first host is yield two times, and because neither of these cases results in a successful read (node2 is stopped), driver uses host with not local rack. I am not sure why that happens.

I changed this slice logic to just iterating through list and the issue stopped reproducing. @Lorak-mmk do you have any idea why?

sylwiaszunejko commented:
@Lorak-mmk ping

Lorak-mmk commented on tests/integration/standard/test_rack_aware_policy.py:
```diff
@@ -0,0 +1,89 @@
+            """)
+
+        for i in range (10):
+            bound = prepared.bind([(i)])
```
What are the inner parenthesis for? They don't create a tuple

sylwiaszunejko replied:
they are there by mistake, I will remove them

sylwiaszunejko commented:
It turns out that using `islice` and `cycle` combination results in broken iterator when there are concurent modification to the list, I checked it using this simple program:
```python
import threading
import time
from itertools import cycle, islice

# Shared list among threads
local_rack_live = ['host1', 'host2', 'host3', 'host4', 'host5']

def remove_elements():
    """Thread function to remove elements from the list."""
    while local_rack_live:
        time.sleep(1)  # Simulate some delay
        removed_host = local_rack_live.pop(0)
        print(f"Removed: {removed_host}")

def iterate_and_yield():
    """Thread function to iterate through the list and yield elements."""
    pos = 0
    while local_rack_live:
        # Ensure pos is valid
        pos = (pos % len(local_rack_live)) if local_rack_live else 0
        for host in islice(cycle(local_rack_live), pos, pos + len(local_rack_live)):
            print(f"Yielding from list: {local_rack_live} host: {host}")
            time.sleep(1)  # Simulate some delay
        pos += 1

# Create threads
remove_thread = threading.Thread(target=remove_elements)
iterate_thread = threading.Thread(target=iterate_and_yield)

# Start threads
remove_thread.start()
iterate_thread.start()

# Wait for threads to complete
remove_thread.join()
iterate_thread.join()
```
The output looks like that:
```
Yielding from list: ['host1', 'host2', 'host3', 'host4', 'host5'] host: host1
Removed: host1
Yielding from list: ['host2', 'host3', 'host4', 'host5'] host: host3
Yielding from list: ['host2', 'host3', 'host4', 'host5'] host: host4
Removed: host2
Removed: host3
Yielding from list: ['host4', 'host5'] host: host1
Yielding from list: ['host4', 'host5'] host: host3
Removed: host4
Yielding from list: ['host5'] host: host5
Removed: host5
```
Clearly it doesn't work returning e.g. host1 when the list looks like [host4, host5].
Because of that we need to use copy of `local_rack_live` to make sure that everything is correct. In other policies there is the same `islice` and `cycle` logic and it is similarly broken; I will address that in a separate PR.

sylwiaszunejko commented:
>  In other policies there is the same islice and cycle logic and it is similarly broken; I will address that in a separate PR.

I have to correct myself here. In other policies such as DCAwareRoundRobinPolicy hosts are stored in a tuple and when they are changes (host up/down) the value in the directory is reassigned so the tuple we get from the dict before the change is unaffected e.g. here: https://github.com/scylladb/python-driver/blob/master/cassandra/policies.py#L272.
So my change from tuples to lists (and changing the list not creating a new one when host down/up) created a problem with the logic similar to other policies.

Lorak-mmk commented:
> > In other policies there is the same islice and cycle logic and it is similarly broken; I will address that in a separate PR.
> 
> I have to correct myself here. In other policies such as DCAwareRoundRobinPolicy hosts are stored in a tuple and when they are changes (host up/down) the value in the directory is reassigned so the tuple we get from the dict before the change is unaffected e.g. here: https://github.com/scylladb/python-driver/blob/master/cassandra/policies.py#L272. So my change from tuples to lists (and changing the list not creating a new one when host down/up) created a problem with the logic similar to other policies.

This approach makes sense - copy in on_up / on_down because those are rare so can be expensive, don't copy in make_query_plan because we want it to be fast.
Maybe thats why they used tuples - to prevent modification.

sylwiaszunejko commented:
> Maybe thats why they used tuples - to prevent accidental modification.

Maybe but it was not that easy to deduct from the code before. So I guess the optimal solution will be to return to using tuples and copy in on_up / on_down, do you agree? @Lorak-mmk

sylwiaszunejko commented on cassandra/policies.py:
```diff
@@ -316,6 +327,112 @@ def on_add(self, host):
+        for host in local_rack_live:
+            yield host
+
+        local_live = [host for host in self._dc_live_hosts.get(self.local_dc, []) if host.rack != self.local_rack]
```
if we return to the version with tuples there will be no need for copying here

Lorak-mmk commented:
> > Maybe thats why they used tuples - to prevent accidental modification.
> 
> Maybe but it was not that easy to deduct from the code before. So I guess the optimal solution will be to return to using tuples and copy in on_up / on_down, do you agree? @Lorak-mmk

I agree. Please also verify that other policies are implemented correctly in this regard. 
It would also be beneficial to put a comment explaining all of this, maybe at the top of the file, so that the next person looking at this has easier time understanding.

Lorak-mmk commented on cassandra/policies.py:
```diff
@@ -316,6 +332,132 @@ def on_add(self, host):
+            if not dc_hosts:
+                return HostDistance.IGNORED
+
+            if host in dc_hosts[:self.used_hosts_per_remote_dc]:
```
IIUC slicing makes a copy of the list. I think it would be better to search for the host in the whole list to avoid unnecessary coopying (`.index()` method).

Lorak-mmk commented on cassandra/policies.py:
```diff
@@ -316,6 +332,132 @@ def on_add(self, host):
+        rack = self._rack(host)
+        with self._hosts_lock:
+            current_rack_hosts = self._live_hosts.get((dc, rack), ())
+            if host not in current_rack_hosts:
+                self._live_hosts[(dc, rack)] = current_rack_hosts + (host, )
+            current_dc_hosts = self._dc_live_hosts.get(dc, ())
+            if host not in current_dc_hosts:
+                self._dc_live_hosts[dc] = current_dc_hosts + (host, )
```
I see that DCAwarePolicy does the same thing. I wonder if it is possible for `on_up` to be marked on already present host. It probably is, otherwise those checks wouldn't exist. Why would driver do this I have no idea.

sylwiaszunejko replied:
I am not sure to be honest
